docs/DevGuide/Brigand.md

```diff
@@ -0,0 +1,1827 @@
+similar type.
+\snippet Test_TMPLDocumentation.cpp tmpl::at
+
+\par
+This operator is \ref map_at "overloaded for maps".
+
+\see \ref at_c
+
+
+\subsubsection at_c
+
+\par
+Retrieves a given element of a list, similar to `operator[]` of the STL
+containers.  The index number is supplied as an `unsigned int`.
+\snippet Test_TMPLDocumentation.cpp tmpl::at_c
+
+\see \ref at
+
+
+\subsubsection back
+
+\par
+Retrieves the last element of a list.
+\snippet Test_TMPLDocumentation.cpp tmpl::back
+
+
+\subsubsection count_if
+
+\par
+Returns the number of elements of a list satisfying a predicate.
+\snippet Test_TMPLDocumentation.cpp tmpl::count_if
+
+\note
+If the predicate is neither a \ref bind "bind expression" with a single
+argument of `_1` nor a \ref metalambda_lazy "lazy expression" with a single
+argument of `_1`, then a bug causes the `type` of the result of the predicate
+to be used instead of the result itself.  As long as the predicate returns an
+\ref integral_constant or a std::integral_constant this does not matter, as
+`type` is a no-op for those classes.
+\snippet Test_TMPLDocumentation.cpp tmpl::count_if:bug:definitions
+\snippet Test_TMPLDocumentation.cpp tmpl::count_if:bug:asserts
+
+
+\subsubsection fold
+
+\par
+Performs a
+[left fold](https://en.wikipedia.org/wiki/Fold_(higher-order_function)), i.e.,
+given a list, initial state, and \ref metalambdas "metalambda", updates the
+state by calling the metalambda on the state and the first element, repeats
+with the second, and so on, returning the final state.
+\snippet Test_TMPLDocumentation.cpp tmpl::fold
+
+\par
+Brigand provides `tmpl::_state` and `tmpl::_element` aliases to the appropriate
+\ref args "arguments" for use in folds.
+
+\see \ref reverse_fold
+
+
+\subsubsection found
+
+\par
+Returns, as an \ref integral_constant of `bool`, whether a predicate matches
+any element of a list.  The default predicate checks that the element's `value`
+is not equal to zero.
+\snippet Test_TMPLDocumentation.cpp tmpl::found
+
+\remark
+This function performs the same operation as \ref any.  See \ref any for
+discussion.
+
+\see \ref any, \ref find, \ref not_found
+
+
+\subsubsection front
+
+\par
+Retrieves the first element of a list.
+\snippet Test_TMPLDocumentation.cpp tmpl::front
+
+
+\subsubsection index_if
+
+\par
+Finds the index as a `size_t` \ref integral_constant of the first type in a
+list satisfying a \ref metalambdas "predicate".  Returns a supplied type,
+defaulting to \ref no_such_type_ if no elements match.
+\snippet Test_TMPLDocumentation.cpp tmpl::index_if
+
+
+\subsubsection index_of
+
+\par
+Finds the index as a `size_t` \ref integral_constant of the first occurrence of
+a type in a list.  Returns \ref no_such_type_ if the type is not found.
+\snippet Test_TMPLDocumentation.cpp tmpl::index_of
+
+
+\subsubsection list_contains
+
+\par
+Checks whether a particular type is contained in a list, returning an
+\ref integral_constant of `bool`.
```

Comment:
When should one choose to return `tmpl::integral_constant<bool,...>` versus `tmpl::true_type`/`tmpl::false_type` versus `std::true_type/std::false_type`?

Reply:
My recommendations:
* If you are writing a metafunction implemented by returning the result of another metafunction, don't do any conversion but directly return whichever of `std` or `tmpl` the wrapped thing returns.
* If you are creating boolean constants yourself:
  * Use the `std` versions unless comparing type equality with the result of something that returns `tmpl` versions.
  * Use `true_type` or `false_type` if you want a literal true or false, and `integral_constant` (or `std::bool_constant` or `tmpl::bool_`) if you want to create it from an expression.

This particular function is implemented in SpECTRE but falls under the first bullet.